### PR TITLE
ST: Check for events

### PR DESCRIPTION
**Pull request by serrss**
### Type of change

- Enhancement / new feature

### Description

Added ability to check events for resources.
It can be useful in case when test deletes Pod and we should be sure that Pod was deleted successfully.

### Checklist

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation

### CODE REVIEW

File: systemtest/src/test/java/io/strimzi/systemtest/KafkaClusterTest.java [hunk 1]
```diff
@@ -143,7 +159,13 @@ public void testKafkaScaleUpScaleDown() {
         for (int brokerId = 0; brokerId < scaleTo; brokerId++) {
             assertTrue(versions, versions.indexOf("(id: " + brokerId + " rack: ") >= 0);
         }
-        // TODO Check for k8s events, logs for errors
+
+        //Test that the new pod does not have errors or failures in events
+        List<Event> events = getEvents("Pod", newPodName);
+        assertThat(events, hasReasons(Scheduled, Pulling, Pulled, Created, Started));
+        assertThat(events, not(hasReasons(Failed, Unhealthy, FailedSync, FailedValidation)));
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/KafkaClusterTest.java]: `not(hasReasons(...))` is checking that `events` *does not have **all** of* those reasons, which is not the same as checking that it has *none of* those reasons. I think you probably need two matchers: `hasAllOf()` and `hasNoneOf()`

**Comment by serrss** [comment]
@tombentley finally, I created two matchers `hasAllOfReasons()` and `hasAnyOfReasons()` that can be used for both cases, when we should check full match and when we should check partial match _(an event has at least one expected reason)_. if we need to check opposite, we can use a construction like `not(hasAnyOfReasons())`.

File: systemtest/src/main/java/io/strimzi/systemtest/matchers/HasAllOfReasons.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        List<String> expectedReasons = Arrays.stream(eventReasons)
+                .map(Enum::name)
+                .collect(Collectors.toList());
+
+        return actualReasons.containsAll(expectedReasons);
+    }
+
+    @Override
+    public void describeTo(Description description) {
+        description.appendValueList("The resource should contain the following event {", ", ", "}. ", eventReasons);
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/matchers/HasAllOfReasons.java]: "...should contain all of the following events"

File: systemtest/src/main/java/io/strimzi/systemtest/matchers/Matchers.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        return new HasAllOfReasons(eventReasons);
+    }
+
+    /**
+     * A matcher checks that examined object has at least one match of reasons for actual events.
+     * @param eventReasons - expected events for resource
+     */
+    public static Matcher<List<Event>> hasAnyOfReasons(Events... eventReasons) {
+        return new HasAnyOfReasons(eventReasons);
+    }
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/matchers/Matchers.java]: I would be inclined to add a `hasNoneOfReasons` method: `return not(hasAnyOfReason(...));`. Doing it like that will make the use sites simpler.

**Comment by serrss** [comment]
@tombentley Added method `hasNoneOfReasons` that will make the use custom matcher simpler.
Thanks for the suggestion

**Pull request merged**